### PR TITLE
Fix test_recurring_jobs_allow_detached_volume & add more recurring job test cases

### PR DESCRIPTION
**Pull request by c3y1huang**
https://github.com/longhorn/longhorn/issues/3517
https://github.com/longhorn/longhorn/issues/3470

### CODE REVIEW

**Review by cchien816 — APPROVED** [review]

File: manager/integration/tests/common.py [hunk 1]
```diff
@@ -2517,7 +2517,6 @@ def wait_pod_auto_attach_after_first_backup_completion(
                 completed = True
                 break
         if completed:
-            wait_for_volume_detached(client, vol.name)
```
Comment (shuo-wu) [inline manager/integration/tests/common.py:2520]: Correct. We should not wait and check a transient state.
But the naming of this function is confusing. In Longhorn, auto attachment typically means that the volume is attached by Longhorn itself. But here the volume attachment is requested by the workload.

File: manager/integration/tests/test_recurring_job.py [hunk 2]
```diff
@@ -1239,100 +1224,199 @@ def test_recurring_job_delete(client, batch_v1_beta_api, volume_name):  # NOQA
...
-    # Should be able to delete recurring job while existing in volume label
+
+@pytest.mark.recurring_job  # NOQA
+def test_recurring_job_delete_should_remove_volume_label(client, batch_v1_beta_api, volume_name):  # NOQA
+    """
+    Scenario: test delete recurring job should remove volume labels
+
+    Given 1 volume created.
+    And  create `snapshot1` recurring job.
+         create `backup1`   recurring job.
```
Comment (shuo-wu) [inline manager/integration/tests/test_recurring_job.py:1235]: And apply these 2 recurring jobs to the volume.

File: manager/integration/tests/test_recurring_job.py [hunk 3]
```diff
@@ -1239,100 +1224,199 @@ def test_recurring_job_delete(client, batch_v1_beta_api, volume_name):  # NOQA
...
+         `backup1`   job     exist in volume recurring job label.
+
+    When delete `backup1` recurring job.
+    Then `snapshot1` job not exist in volume recurring job label.
+         `backup1`   job not exist in volume recurring job label.
+
+    Given create `snapshot1` recurring job with `group-1` in groups.
+    And   create `snapshot2` recurring job with `group-1` in groups.
+    And   create `backup1`   recurring job with `group-1` in groups.
+    And   remove volume recurring job label `default` job-group.
```
Comment (shuo-wu) [inline manager/integration/tests/test_recurring_job.py:1250]: Group `default` will be applied to the volume automatically, but Longhorn won't do this for group `group-1`. Need to clarify this part here.

File: manager/integration/tests/test_recurring_job.py [hunk 4]
```diff
@@ -1239,100 +1224,199 @@ def test_recurring_job_delete(client, batch_v1_beta_api, volume_name):  # NOQA
...
+    wait_for_volume_recurring_job_update(volume,
+                                         jobs=[], groups=[group1])
+
+    snap2_recurring_job = client.by_id_recurring_job(snap2)
+    client.delete(snap2_recurring_job)
+    time.sleep(5)
+    wait_for_volume_recurring_job_update(volume,
+                                         jobs=[], groups=[group1])
+
+    # Delete last recurring job in group remove the volume label
```
Comment (shuo-wu) [inline manager/integration/tests/test_recurring_job.py:1337]: Do you mean `Deleting last recurring job of the group would lead to the group label being cleaned up for all volumes`? We need to add this warning message/tip in Longhorn UI.
Does this logic apply to the default group?

**Review by shuo-wu — CHANGES_REQUESTED** [review]
Is there a test case for the following corner case?
Add/Remove a recurring job for a volume of which the job group already contains the recurring job.

**Comment by kaxing** [comment]
Tried running these changes on my local env and it seems something didn't add up. just leave logs here.

Both `test_recurring_job_delete_should_remove_volume_label` and `test_recurring_job_volume_label_when_job_and_group_use_same_name` are not able to finish without error.

<Details>

<Summary>
Test result logs:
</Summary>

```
root@longhorn-test:/integration/tests# ./run.sh test_recurring_job.py
+ flake8
+ pytest -v test_recurring_job.py
======================================================================================= test session starts =======================================================================================
platform linux -- Python 3.9.10, pytest-5.3.1, py-1.11.0, pluggy-0.13.1 -- /usr/local/bin/python
cachedir: .pytest_cache
rootdir: /integration, inifile: pytest.ini
plugins: repeat-0.9.1
collected 28 items

test_recurring_job.py::test_recurring_job[s3] PASSED                                                                                                                                        [  3%]
test_recurring_job.py::test_recurring_job[nfs] PASSED                                                                                                                                       [  7%]
test_recurring_job.py::test_recurring_job_in_volume_creation PASSED                                                                                                                         [ 10%]
test_recurring_job.py::test_recurring_job_duplicated PASSED                                                                                                                                 [ 14%]
test_recurring_job.py::test_recurring_job_in_storageclass[s3] PASSED                                                                                                                        [ 17%]
test_recurring_job.py::test_recurring_job_in_storageclass[nfs] PASSED                                                                                                                       [ 21%]
test_recurring_job.py::test_recurring_job_labels[s3] PASSED                                                                                                                                 [ 25%]
test_recurring_job.py::test_recurring_job_labels[nfs] PASSED                                                                                                                                [ 28%]
test_recurring_job.py::test_recurring_job_kubernetes_status[s3] PASSED                                                                                                                      [ 32%]
test_recurring_job.py::test_recurring_job_kubernetes_status[nfs] PASSED                                                                                                                     [ 35%]
test_recurring_job.py::test_recurring_jobs_maximum_retain PASSED                                                                                                                            [ 39%]
test_recurring_job.py::test_recurring_job_detached_volume PASSED                                                                                                                            [ 42%]
test_recurring_job.py::test_recurring_jobs_allow_detached_volume[s3] PASSED                                                                                                                 [ 46%]
test_recurring_job.py::test_recurring_jobs_allow_detached_volume[nfs] PASSED                                                                                                                [ 50%]
test_recurring_job.py::test_recurring_jobs_when_volume_detached_unexpectedly[s3] PASSED                                                                                                     [ 53%]
test_recurring_job.py::test_recurring_jobs_when_volume_detached_unexpectedly[nfs] PASSED                                                                                                    [ 57%]
test_recurring_job.py::test_recurring_jobs_on_nodes_with_taints SKIPPED                                                                                                                     [ 60%]
test_recurring_job.py::test_recurring_job_groups[s3] PASSED                                                                                                                                 [ 64%]
test_recurring_job.py::test_recurring_job_groups[nfs] PASSED                                                                                                                                [ 67%]
test_recurring_job.py::test_recurring_job_default PASSED                                                                                                                                    [ 71%]
test_recurring_job.py::test_recurring_job_delete PASSED                                                                                                                                     [ 75%]
test_recurring_job.py::test_recurring_job_delete_should_remove_volume_label FAILED                                                                                                          [ 78%]
test_recurring_job.py::test_recurring_job_volume_label_when_job_and_group_use_same_name FAILED                                                                                              [ 82%]
test_recurring_job.py::test_recurring_job_multiple_volumes[s3] PASSED                                                                                                                       [ 85%]
test_recurring_job.py::test_recurring_job_multiple_volumes[nfs] PASSED                                                                                                                      [ 89%]
test_recurring_job.py::test_recurring_job_snapshot PASSED                                                                                                                                   [ 92%]
test_recurring_job.py::test_recurring_job_backup[s3] PASSED                                                                                                                                 [ 96%]
test_recurring_job.py::test_recurring_job_backup[nfs] PASSED                                                                                                                                [100%]

============================================================================================ FAILURES =============================================================================================
______________________________________________________________________ test_recurring_job_delete_should_remove_volume_label _______________________________________________________________________

client = <longhorn.Client object at 0x7f832e75cee0>, batch_v1_beta_api = <kubernetes.client.api.batch_v1beta1_api.BatchV1beta1Api object at 0x7f832e255040>
volume_name = 'longhorn-testvol-0jmf3u'

    @pytest.mark.recurring_job  # NOQA
    def test_recurring_job_delete_should_remove_volume_label(client, batch_v1_beta_api, volume_name):  # NOQA
        """
        Scenario: test delete recurring job should remove volume labels

        Given 1 volume created.
        And  create `snapshot1` recurring job.
             create `backup1`   recurring job.
        And  `snapshot1` job exist in volume recurring job label.
             `backup1`   job exist in volume recurring job label.

        When delete `snapshot1` recurring job.
        Then `snapshot1` job not exist in volume recurring job label.
             `backup1`   job     exist in volume recurring job label.

        When delete `backup1` recurring job.
        Then `snapshot1` job not exist in volume recurring job label.
             `backup1`   job not exist in volume recurring job label.

        Given create `snapshot1` recurring job with `group-1` in groups.
        And   create `snapshot2` recurring job with `group-1` in groups.
        And   create `backup1`   recurring job with `group-1` in groups.
        And   remove volume recurring job label `default` job-group.
        And  `group-1` job-group     exist in volume recurring job label.
             `default` job-group not exist in volume recurring job label.

        When delete `snapshot1` recurring job.
        Then `group-1` job-group exist in volume recurring job label.

        When delete `snapshot2` recurring job.
        Then `group-1` job-group exist in volume recurring job label.

        When delete `back1` recurring job.
        Then `group-1` job-group not exist in volume recurring job label.
        """
        client.create_volume(name=volume_name, size=SIZE)
        volume = wait_for_volume_detached(client, volume_name)

        snap1 = SNAPSHOT + "1"
        snap2 = SNAPSHOT + "2"
        back1 = BACKUP + "1"
        group1 = "group-1"
        snap_job = {
            TASK: SNAPSHOT,
            GROUPS: [group1],
            CRON: SCHEDULE_1MIN,
            RETAIN: 1,
            CONCURRENCY: 2,
            LABELS: {},
        }
        back_job = {
            TASK: BACKUP,
            GROUPS: [group1],
            CRON: SCHEDULE_1MIN,
            RETAIN: 1,
            CONCURRENCY: 2,
            LABELS: {},
        }
        recurring_jobs = {
            snap1: snap_job,
            back1: back_job,
        }
        create_recurring_jobs(client, recurring_jobs)
        check_recurring_jobs(client, recurring_jobs)

        # Delete recurring job should remove volume label
        volume.recurringJobAdd(name=snap1, isGroup=False)
        volume.recurringJobAdd(name=back1, isGroup=False)
        wait_for_volume_recurring_job_update(volume,
                                             jobs=[snap1, back1], groups=[DEFAULT])

        snap1_recurring_job = client.by_id_recurring_job(snap1)
        back1_recurring_job = client.by_id_recurring_job(back1)
        client.delete(snap1_recurring_job)
>       wait_for_volume_recurring_job_update(volume,
                                             jobs=[back1], groups=[DEFAULT])

test_recurring_job.py:1302:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

volume = {'accessMode': 'rwo', 'backingImage': '', 'backupStatus': [], 'cloneStatus': {'snapshot': '', 'sourceVolume': '', 'sta...reEndpoint': '', 'shareState': '', 'size': '16777216', 'staleReplicaTimeout': 0, 'standby': False, 'state': 'detached'}
jobs = ['backup1'], groups = ['default']

    def wait_for_volume_recurring_job_update(volume, jobs=[], groups=[]):
        ok = False
        for _ in range(RETRY_COUNTS):
            volumeJobs, volumeGroups = get_volume_recurring_jobs_and_groups(volume)
            try:
                assert len(volumeGroups) == len(groups)
                for group in groups:
                    assert group in volumeGroups

                assert len(volumeJobs) == len(jobs)
                for job in jobs:
                    assert job in volumeJobs

                ok = True
                break
            except AssertionError:
                time.sleep(RETRY_INTERVAL)
>       assert ok
E       AssertionError

common.py:4419: AssertionError
________________________________________________________________ test_recurring_job_volume_label_when_job_and_group_use_same_name _________________________________________________________________

client = <longhorn.Client object at 0x7f832e3558b0>, volume_name = 'longhorn-testvol-boeopp'

    @pytest.mark.recurring_job  # NOQA
    def test_recurring_job_volume_label_when_job_and_group_use_same_name(client, volume_name):  # NOQA
        """
        Scenario: test volume label recurring job when recurring job and
                  job-group uses the same name.

        Given volume-1 created.
              volume-2 created.
              volume-3 created.
        And  create `snapshot1` recurring job with `snapshot-1` in groups.
             create `snapshot2` recurring job with `snapshot-1` in groups.
        And  `snapshot1` job-group exist in volume-1 recurring job label.
             `snapshot1` job       exist in volume-2 recurring job label.
             `snapshot2` job       exist in volume-3 recurring job label.

        When delete `snapshot1` recurring job.
        Then `snapshot1` job-group     exist in volume-1 recurring job label.
             `snapshot1` job       not exist in volume-2 recurring job label.
             `snapshot2` job           exist in volume-3 recurring job label.

        When delete `snapshot2` recurring job.
        Then `snapshot1` job-group not exist in volume-1 recurring job label.
             `snapshot2` job       not exist in volume-3 recurring job label.
        """
        volume1_name = volume_name + "-1"
        volume2_name = volume_name + "-2"
        volume3_name = volume_name + "-3"
        client.create_volume(name=volume1_name, size=SIZE)
        client.create_volume(name=volume2_name, size=SIZE)
        client.create_volume(name=volume3_name, size=SIZE)
        volume1 = wait_for_volume_detached(client, volume1_name)
        volume2 = wait_for_volume_detached(client, volume2_name)
        volume3 = wait_for_volume_detached(client, volume3_name)

        snap1 = SNAPSHOT + "1"
        snap2 = SNAPSHOT + "2"
        snapshot = {
            TASK: SNAPSHOT,
            GROUPS: [snap1],
            CRON: SCHEDULE_1MIN,
            RETAIN: 1,
            CONCURRENCY: 2,
            LABELS: {},
        }
        recurring_jobs = {
            snap1: snapshot,
            snap2: snapshot,
        }
        create_recurring_jobs(client, recurring_jobs)
        check_recurring_jobs(client, recurring_jobs)

        # Delete recurring job should correctly remove volume label
        volume1.recurringJobAdd(name=snap1, isGroup=True)
        volume2.recurringJobAdd(name=snap1, isGroup=False)
        volume3.recurringJobAdd(name=snap2, isGroup=False)
        wait_for_volume_recurring_job_update(volume1,
                                             jobs=[], groups=[snap1, DEFAULT])
        wait_for_volume_recurring_job_update(volume2,
                                             jobs=[snap1], groups=[DEFAULT])
        wait_for_volume_recurring_job_update(volume3,
                                             jobs=[snap2], groups=[DEFAULT])

        snap1_recurring_job = client.by_id_recurring_job(snap1)
        snap2_recurring_job = client.by_id_recurring_job(snap2)
        client.delete(snap1_recurring_job)
>       wait_for_volume_recurring_job_update(volume2,
                                             jobs=[], groups=[DEFAULT])

test_recurring_job.py:1409:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

volume = {'accessMode': 'rwo', 'backingImage': '', 'backupStatus': [], 'cloneStatus': {'snapshot': '', 'sourceVolume': '', 'sta...reEndpoint': '', 'shareState': '', 'size': '16777216', 'staleReplicaTimeout': 0, 'standby': False, 'state': 'detached'}
jobs = [], groups = ['default']

    def wait_for_volume_recurring_job_update(volume, jobs=[], groups=[]):
        ok = False
        for _ in range(RETRY_COUNTS):
            volumeJobs, volumeGroups = get_volume_recurring_jobs_and_groups(volume)
            try:
                assert len(volumeGroups) == len(groups)
                for group in groups:
                    assert group in volumeGroups

                assert len(volumeJobs) == len(jobs)
                for job in jobs:
                    assert job in volumeJobs

                ok = True
                break
            except AssertionError:
                time.sleep(RETRY_INTERVAL)
>       assert ok
E       AssertionError

common.py:4419: AssertionError
====================================================================== 2 failed, 25 passed, 1 skipped in 8331.90s (2:18:51) =======================================================================
```
</Details>

**Comment by c3y1huang** [comment]
> Tried running these changes on my local env and it seems something didn't add up. just leave logs here.
> 
> Both `test_recurring_job_delete_should_remove_volume_label` and `test_recurring_job_volume_label_when_job_and_group_use_same_name` are not able to finish without error.

The tests are dependent on this [PR](https://github.com/longhorn/longhorn-manager/pull/1223).

File: manager/integration/tests/common.py (hunk 1 above)
Comment (c3y1huang) [reply manager/integration/tests/common.py:2520]: Renamed.

File: manager/integration/tests/test_recurring_job.py (hunk 2 above)
Comment (c3y1huang) [reply manager/integration/tests/test_recurring_job.py:1235]: Addressed.

File: manager/integration/tests/test_recurring_job.py (hunk 3 above)
Comment (c3y1huang) [reply manager/integration/tests/test_recurring_job.py:1256]: Added comment for clarification.

File: manager/integration/tests/test_recurring_job.py (hunk 4 above)
Comment (c3y1huang) [reply manager/integration/tests/test_recurring_job.py]: > Do you mean `Deleting last recurring job of the group would lead to the group label being cleaned up for all volumes`?

Reworded.

> We need to add this warning message/tip in Longhorn UI. 
Sure I will create an issue.

>Does this logic apply to the default group?
Should not affect the default group auto-assignment. I've added steps to the test case.

Comment (c3y1huang) [reply manager/integration/tests/test_recurring_job.py]: https://github.com/longhorn/longhorn/issues/3651

**Comment by c3y1huang** [comment]
> Is there a test case for the following corner case? Add/Remove a recurring job for a volume of which the job group already contains the recurring job.

It's covered in `test_recurring_job_volume_label_when_job_and_group_use_same_name`.

File: manager/integration/tests/common.py [hunk 5]
```diff
@@ -2505,7 +2505,7 @@ def wait_for_backup_completion(client, volume_name, snapshot_name=None,
     return v
 
 
-def wait_pod_auto_attach_after_first_backup_completion(
+def wait_pod_attach_after_first_backup_completion(
```
Comment (shuo-wu) [inline manager/integration/tests/common.py:2508]: NIT: This renaming should be in the 1st commit

File: manager/integration/tests/test_recurring_job.py (hunk 3 above)
Comment (shuo-wu) [reply manager/integration/tests/test_recurring_job.py:1256]: You haven't mentioned that `group-1` is added to the volume

File: manager/integration/tests/test_recurring_job.py [hunk 6]
```diff
@@ -1239,66 +1224,107 @@ def test_recurring_job_delete(client, batch_v1_beta_api, volume_name):  # NOQA
...
+    Given create `snapshot1` recurring job with `group-1` in groups.
+    And   create `snapshot2` recurring job with `group-1` in groups.
+    And   create `backup1`   recurring job with `group-1` in groups.
+    And   create `backup2`   recurring job with `default` in groups.
+    // The default job-group automatically applies to the volumes
+    // with no recurring job. We want to keep the test focused on the
+    // behavior so removing the default job-group assignment first.
+    And   remove volume recurring job label `default` job-group.
+    And  `group-1` job-group     exist in volume recurring job label.
+         `default` job-group not exist in volume recurring job label.
```
Comment (shuo-wu) [inline manager/integration/tests/test_recurring_job.py:1258]: Will group `default` be re-added to the volume automatically? If YES, you should not check this.

**Review by shuo-wu — CHANGES_REQUESTED** [review]

File: manager/integration/tests/common.py (hunk 5 above)
Comment (c3y1huang) [reply manager/integration/tests/common.py:2508]: Addressed.

File: manager/integration/tests/test_recurring_job.py (hunk 3 above)
Comment (c3y1huang) [reply manager/integration/tests/test_recurring_job.py:1256]: Fixed.

File: manager/integration/tests/test_recurring_job.py (hunk 6 above)
Comment (shuo-wu) [reply manager/integration/tests/test_recurring_job.py:1259]: OK. Then please add this explanation here. And maybe we need to mention users as well when they try to remove group `default` from the volume. I already forgot this behavior...

**Review by shuo-wu — APPROVED** [review]
In general LGTM

File: manager/integration/tests/test_recurring_job.py (hunk 6 above)
Comment (c3y1huang) [reply manager/integration/tests/test_recurring_job.py:1259]: This is mentioned in [doc](https://longhorn.io/docs/1.2.3/snapshots-and-backups/scheduling-backups-and-snapshots/#set-up-recurring-jobs-to-be-the-default).

**Review by innobead — APPROVED** [review]
